### PR TITLE
Converter for LogicalNot operation

### PR DESCRIPTION
**Pull request by drivanov**
- Converter for LogicalNot operation is using the same base class (`ConvertUnaryImpl`) as other Unary operations. 
- The special converter for `Rsqrt
` was removed and now for this operation regular Unary Op converter is used.
- New templated class implemented for testing `ConvertUnary`, `ConvertBooleanUnary`, `ConvertActivation`.
- A new check and corresponding subtests were added to Validate: at least 1 dimension is required for input of any `Unary` and `UnaryBoolean` operation. (Similar subtest for `ConvertActivation` operations is blocked and after refactoring of `ConvertActivation` it will be activated)

### CODE REVIEW

**Comment by drivanov** [comment]
Should be merged AFTER [PR#55177](https://github.com/tensorflow/tensorflow/pull/55177): **Update test helpers to handle bool tensors and weights**

**Review by bishalbaaniya123 — APPROVED** [review]

**Comment by drivanov** [comment]
I fixed some formatting issues and noticed that this PR was closed.

**Comment by bixia1** [comment]
Would you please rebase this PR? @drivanov

**Comment by drivanov** [comment]
> Would you please rebase this PR? @drivanov

Rebase is done. 

@bixia1 : Just one note... On Friday (03/18/2022) I noticed that my latest change for [PR#55177](https://github.com/tensorflow/tensorflow/pull/55177/) fixes the issue with `ConvertMatMul` you mentioned, but is not valid for `LogicalNot` and other boolean operations. I fixed this and updated [PR#55177](https://github.com/tensorflow/tensorflow/pull/55177/), but it seems that it was already merged at that point. After that, I got the "merge conflict" messages for all my 4 PRs and I fixed them yesterday (03/20/22).

Long story short, the changes in [that fragments](https://github.com/tensorflow/tensorflow/blob/516c77c4c78770ae246b6ff7255c6e6f5b6674d7/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L1374-L1435) is a correct fix for the the issue with `ConvertMatMul` you mentioned.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc [hunk 1]
```diff
@@ -1848,6 +1838,105 @@ class ParameterizedOpConverterTestBase
...
+          (op_name == "Sign" || op_name == "Round" ||
+           op_name == "LogicalNot")) {
+        conv_status =
+            errors::Unimplemented("Unary op: '", op_name,
+                                  "' is not supported in implicit batch mode");
+      }
+
+      Reset();
+      // TODO(bixia): we assume this test is only instantiated for DT_FLOAT for
+      // now. Need to find a better way to express input and output types.
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1887]: Is this TODO still true, as we now instantiate the test for bool?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc [hunk 2]
```diff
@@ -1848,6 +1838,105 @@ class ParameterizedOpConverterTestBase
...
+      //
+      // TODO(tfeher): improve tests by defining an expected output data type
+      //  and check that. Currently only the shape and values of the output are
+      // checked.
+      const DataType input_tf_type = op_name == "Cast" ? DT_HALF : tf_type;
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1892]: Would this comment be proper for this line:
For CastOp, tf_type is the result type and we use DT_HALF as the input type.

Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1892]: Currently cast only tests TD_HALF -> DT_FLOAT conversion. https://github.com/tensorflow/tensorflow/blob/24c3d0a72f534944de8505ccf13103534378469e/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L6906-L6907

A suggest to remove the comment, and add the following:
```suggestion
      const DataType input_tf_type = op_name == "Cast" ? DT_HALF : tf_type;
      const DataType output_tf_type = op_name == "Cast" ? DT_FLOAT : tf_type;
```
The `out_tf_type` variable shall be used while calling TestOpconverter.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc [hunk 3]
```diff
@@ -1848,6 +1838,105 @@ class ParameterizedOpConverterTestBase
...
+
+      AddTestTensor("input", p.input_dims, input_tf_type, input_values);
+
+      std::vector<float> output;
+      std::transform(input_values.begin(), input_values.end(),
+                     std::back_inserter(output), op_map[op_name].second);
+
+      TestOpConverter("my_unary", node_def, p.expected_output_dims, conv_status,
+                      runtime_status,
+                      ArrayFloatNear(output, max_abs_error, nan_sensitive));
```
Comment (tfeher) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1902]:
```suggestion
      TestOpConverter("my_unary", node_def, p.expected_output_dims, conv_status,
                      runtime_status,
                      ArrayFloatNear(output, max_abs_error, nan_sensitive), {output_tf_type});
```

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc [hunk 4]
```diff
@@ -1848,6 +1838,105 @@ class ParameterizedOpConverterTestBase
...
+
+      if (checkOutput) {
+        TRT_TensorOrWeights output;
+        TF_EXPECT_OK(GetTensorOrWeights("my_unary", &output));
+      }
```
Comment (tfeher) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1907]: Why do we need this? Does this catch any error not seen by `TestOpConverter`?

**Review by tfeher — CHANGES_REQUESTED** [review]
Hi Andrei, thanks for the PR! Just a few small questions / suggestions.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc (hunk 4 above)
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1907]: @tfeher : Sorry, I have no idea why do we need this. When implementing 
```
template <typename T>
class OpConverter_UnaryTest : public ParameterizedOpConverterTestBase {
``` 
which is used for `ConvertUnary`, `ConvertBoolUnary` and `ConvertActivation` tests, I just copied [these lines](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L4098-L4099) from the old version of 
```
TEST_P(OpConverter_FP32_Test, ConvertActivation) {
```
and made them active only for `ConvertActivation` test, when `checkOutput` is TRUE.

**Do you want me to delete them?**

Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1903]: Please remove this then. `TestOpConverter` does already similar checks while retrieving the output values.

**Review by tfeher — APPROVED** [review]
Thanks Andrei for the updates, it looks good to me.

**Review by bixia1 — APPROVED** [review]
